src/index.ts

```diff
@@ -176,19 +96,118 @@ function compileToStringSync(sources, options) {
+}
+
+function runCompiler(sources: string | string[], options: Options, pathToElm: string): ChildProcess {
+  if (typeof options.spawn !== "function") {
+    throw "options.spawn was a(n) " + (typeof options.spawn) + " instead of a function.";
+  }
+
+  var processArgs = prepareProcessArgs(sources, options);
+  var processOpts = prepareProcessOpts(options);
+
+  if (options.verbose) {
+    console.log(["Running", pathToElm].concat(processArgs).join(" "));
+  }
+
+  return options.spawn(pathToElm, processArgs, processOpts);
+}
+
+function prepareProcessArgs(sources: string | string[], options: Options): string[] {
+  var preparedSources = prepareSources(sources);
+  var compilerArgs = compilerArgsFromOptions(options);
+
+  return ["make"].concat(preparedSources ? preparedSources.concat(compilerArgs) : compilerArgs);
+}
+
+function prepareSources(sources: string | string[]): string[] {
+  if (!(sources instanceof Array || typeof sources === "string")) {
+    throw "compile() received neither an Array nor a String for its sources argument.";
+  }
+
+  return typeof sources === "string" ? [sources] : sources;
+}
+
+function prepareProcessOpts(options: Options): SpawnOptions {
+  var env = _.merge({ LANG: 'en_US.UTF-8' }, process.env);
+  return _.merge({ env: env, stdio: "inherit", cwd: options.cwd }, options.processOpts);
+
+}
+
+function compilerErrorToString(err: { code?: string, message?: string }, pathToElm: string): string {
+  if ((typeof err === "object") && (typeof err.code === "string")) {
+    switch (err.code) {
+      case "ENOENT":
+        return "Could not find Elm compiler \"" + pathToElm + "\". Is it installed?";
+
+      case "EACCES":
+        return "Elm compiler \"" + pathToElm + "\" did not have permission to run. Do you need to give it executable permissions?";
+
+      default:
+        return "Error attempting to run Elm compiler \"" + pathToElm + "\":\n" + err;
+    }
+  } else if ((typeof err === "object") && (typeof err.message === "string")) {
+    return JSON.stringify(err.message);
+  } else {
+    return "Exception thrown when attempting to run Elm compiler " + JSON.stringify(pathToElm);
+  }
+}
+
+function getSuffix(outputPath: string | undefined, defaultSuffix: string): string {
+  if (outputPath) {
+    return path.extname(outputPath) || defaultSuffix;
+  } else {
+    return defaultSuffix;
+  }
+}
+
 // Converts an object of key/value pairs to an array of arguments suitable
 // to be passed to child_process.spawn for elm-make.
-function compilerArgsFromOptions(options) {
-  return _.flatten(_.map(options, function (value, opt) {
+function compilerArgsFromOptions(options: Options): string[] {
+  return _.flatten(_.map(options, function (value, opt): string[] {
     if (value) {
       switch (opt) {
         case "help": return ["--help"];
-        case "output": return ["--output", value];
-        case "report": return ["--report", value];
+        case "output": return ["--output", (value as string)];
```

Comment:
can we not type `value` in the function definition? `function (value: string, opt)` or does that not work? Sorry if I've misunderstood!